api/pom.xml

```diff
@@ -103,23 +103,52 @@
 
     <properties>
         <api_package>javax.websocket</api_package>
-        <spec_version>1.1</spec_version>
+        <spec_version>1.2</spec_version>
         <!--<spec_build>02</spec_build>-->
-        <new_spec_version>1.1</new_spec_version>
-        <spec_impl_version>1.1.1</spec_impl_version>
-        <packages.export>javax.websocket.*</packages.export>
+        <new_spec_version>1.2</new_spec_version>
+        <spec_impl_version>1.2.1</spec_impl_version>
     </properties>
 
     <build>
         <plugins>
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
-                <version>2.5.1</version>
+                <version>3.8.0</version>
                 <configuration>
-                    <source>1.6</source>
-                    <target>1.6</target>
+                    <jdkToolchain>
+                        <version>9</version>
+                    </jdkToolchain>
+                    <!-- Enforce low bytecode version to preserve compatibility -->
+                    <release>6</release>
+                    <!-- Set source-target language levels to ENABLE use of module-info in the first place, even if bytecode is of lower level -->
+                    <source>9</source>
```

Comment:
Changing the compiled bytecode to java 9+ is undesirable.
Are there any other Jakarata EE projects with a module-info.class yet?
(Last I heard, only the automatic module name was used)

Reply:
I sampled a few class files for bytecode version, it's still 50.0 (because of the `<release>6</release>` element. Of course, `module-info.class` is bytecode level 9, but I don't think that matters for older VMs since they won't be reading that class in any case.

I know that it doesn't really matter in another organization, but personally, I have used this pattern to create a JPMS-compliant test helper library before that happens to be usable under Java 8 as well.

Reply:
In that case, shouldn't the `module-info.class` be in the JEP238 compatible `META-INF/versions/9/module-info.class` so as to not be exposed to Java 8 Runtime environments (such as those that perform bytecode scanning for annotations)

Reply:
I've been researching this, and while I agree that that would be the _ideal_ way of going about it, I have yet to find a way that preserves _both_ the OSGi bundle structure _and_ allows the compiler to add in the `module-info` class automatically.

To expand, the Maven compiler plugin has a specific configuration for multi-release JARs, but that places the classes in their respective version-specific directories, which, as far as I know, does not satisfy the OSGi spec. It _would_ hide the Java 9 bytecode from scanning, though.
On the other hand, the `buildDirectory` configuration of the compiler plugin, which _could_ be a way to preserve the OSGi structure, seems to be ignored, so I can't really do a different sort of two-pass compilation (because even this solution is two-pass), unless I _massively_ reorganize the source tree to create additional Maven modules and their related build processes...

Reply:
Don't worry about OSGi, it **must** support JEP238, otherwise many things would fail today.

Reply:
Well, if OSGi is not a deal-breaker, things are much simpler! I was trying to preserve as much of the current compatibility as possible.

That said, I'm having my doubts now, because Maven doesn't seem to find the required modules under the JEP238 architecture. I'm waiting for the results of the Travis build now, see if a clean environment makes a difference.

Reply:
Well, I got the JDK8 build to pass as well. I'm not sure why the legal check is failing, I _have_ signed the ECA...

Reply:
Regarding the legal check ...

> ERROR: Zalan Meggyesi (zmeggyesi@****.hu) did not include the "Signed-off-by footer" which is required for all commits made by a contributor.

You have 6 commits without that footer.

Reply:
Right...
And since I can't edit already-pushed commits, I think I'll close this PR,
rebase my commits locally, editing them to include the footer, and open a
new one. If you think the code itself is fit to merge, that is...

Reply:
No need to close / reopen.
Just rebase and push to your branch with `git push origin --force-with-lease websocket-api`
That will update the entire PR with your current rebase.
Heck, squash them all into a single commit if you want. (probably a good idea)

Reply:
I'm normally quite averse to rebasing (my organization discourages it), but it seems to have gone well in this case.

Thank you for all your input and assistance in this change - it was very much appreciated!